gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java

```diff
@@ -1795,31 +1795,37 @@ public List<GbStudentGradeInfo> sortGradeMatrix(Map<String, GbStudentGradeInfo>
 	 */
 	public List<GbGroup> getSiteSectionsAndGroups() {
 		final String siteId = getCurrentSiteId();
+		final String userId = getCurrentUser().getId();
 
 		final List<GbGroup> rval = new ArrayList<>();
 
-		GbRole role;
-		try {
-			role = this.getUserRole(siteId);
-		} catch (final GbAccessDeniedException e) {
-			log.warn("Could not fetch the users role in site [{}], {}", siteId, e.toString());
-			return rval;
-		}
-
 		// get groups (handles both groups and sections)
 		try {
-			final Site site = this.siteService.getSite(siteId);
-			final Collection<Group> groups = isSuperUser() || role == GbRole.INSTRUCTOR ? site.getGroups() : site.getGroupsWithMember(userDirectoryService.getCurrentUser().getId());
+			
+		final Site site = this.siteService.getSite(siteId);
 
-			for (final Group group : groups) {
-				rval.add(new GbGroup(group.getId(), group.getTitle(), group.getReference(), GbGroup.Type.GROUP));
-			}
+		final List<PermissionDefinition> perms = getPermissionsForUser(userId, siteId);
+
+		List<String> groupReferences = perms.stream().map(PermissionDefinition::getGroupReference).collect(Collectors.toList());
+
+		final Collection<Group> groups = !groupReferences.isEmpty() ? site.getGroups() : site.getGroupsWithMember(userDirectoryService.getCurrentUser().getId());
```

Comment:
If groupReferences is not empty you are returning all the groups for the site. What if the user only has permission on a subset of the site's groups?